File: 프로그래머스/2/77885. 2개 이하로 다른 비트/2개 이하로 다른 비트.js
function solution(numbers) {
    return numbers.map((n) => {
        // 2진수 변환
        let str = '0' + n.toString(2);
        if(str[str.length-1] === '0')
            str = str.substring(0, str.length-1) + '1';
        else{
            let idx = str.lastIndexOf('01');
            // 01을 10으로 바꾸는게 해당 문제의 값을 구하는 베스트
            str = str.substring(0, idx) + '10' + str.substring(idx+2, str.length);
        }
        
        return parseInt(str, 2);
    });
}

File: 프로그래머스/2/77885. 2개 이하로 다른 비트/2개 이하로 다른 비트.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

vector<long long> solution(vector<long long> numbers) {
    vector<long long> answer;
    string str = "";
    
    for(auto n : numbers){
        str = "";
        
        if(n == 0){
            answer.push_back(1);
            continue;
        }
        
        while(n > 0){
            str += to_string(n%2);
            n /= 2;
        }
        
        long long cnt=0;
        for(int i=0; i<str.length(); i++){
            // 하단에서 0을 1로 바꾸는게 베스트(reverse 기준)
            if(str[i] == '0'){
                str[i] = '1';
                cnt = i-1;
                break;
            }
            
            // 0이 없을 경우 상단에 1추가(reverse 기준)
            if(i == str.length()-1){
                str += '1';
                cnt = i;
                break;
            }
        }
        
        for(int i=cnt; i>=0; i--){
            if(str[i] == '1'){
                str[i] = '0';
                break;
            }
        }
        
        long long sum = 0;
        for(int i=0; i<str.length(); i++)
            sum += (str[i]-'0') * pow(2, i);
        
        answer.push_back(sum);
    }
    
    return answer;
}